adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: [ %d ]", response.StatusCode),
+		}
+	}
+
+	if response.StatusCode == http.StatusServiceUnavailable {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Something went wrong, please contact your Account Manager. Status Code: [ %d ] ", response.StatusCode),
+		}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: [ %d ]. Run with request.debug = 1 for more info", response.StatusCode),
+		}
+	}
+
+	if response.Body == nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("bidderRawResponse body is empty"),
+		}
+	}
+	return nil
+}
+
+// check HuaweiAds response retcode
+func checkHuaweiAdsResponseRetcode(response huaweiAdsResponse) error {
+	if response.Retcode == 200 || response.Retcode == 204 || response.Retcode == 206 {
+		return nil
+	}
+
+	if (response.Retcode < 600 && response.Retcode >= 400) || (response.Retcode < 300 && response.Retcode > 200) {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("HuaweiAdsResponse retcode: %d , reason: %s", response.Retcode, response.Reason),
+		}
+	}
+	return nil
+}
+
+// convert HuaweiAds' response into bidder's response
+func (a *adapter) convertHuaweiAdsResp2BidderResp(huaweiAdsResponse *huaweiAdsResponse, openRTBRequest *openrtb2.BidRequest) (bidderResponse *adapters.BidderResponse, err error) {
+	if len(huaweiAdsResponse.Multiad) == 0 {
+		return nil, errors.New("convertHuaweiAdsResp2BidderResp: multiad length is 0, get no ads from huawei side.")
+	}
+	bidderResponse = adapters.NewBidderResponseWithBidsCapacity(len(huaweiAdsResponse.Multiad))
+	// Default Currency: CNY
+	bidderResponse.Currency = "CNY"
+
+	// record request Imp (slotid->imp, slotid->openrtb_ext.bidtype)
+	mapSlotid2Imp := make(map[string]openrtb2.Imp, len(openRTBRequest.Imp))
+	mapSlotid2MediaType := make(map[string]openrtb_ext.BidType, len(openRTBRequest.Imp))
+	for _, imp := range openRTBRequest.Imp {
+		huaweiAdsExt, err := unmarshalExtImpHuaweiAds(&imp)
+		if err != nil {
+			continue
+		}
+		mapSlotid2Imp[huaweiAdsExt.SlotId] = imp
+
+		var mediaType = openrtb_ext.BidTypeBanner
+		if imp.Video != nil {
+			mediaType = openrtb_ext.BidTypeVideo
+		} else if imp.Native != nil {
+			mediaType = openrtb_ext.BidTypeNative
+		} else if imp.Audio != nil {
+			mediaType = openrtb_ext.BidTypeAudio
+		}
+		mapSlotid2MediaType[huaweiAdsExt.SlotId] = mediaType
+	}
+
+	if len(mapSlotid2MediaType) < 1 || len(mapSlotid2Imp) < 1 {
+		return nil, errors.New("convertHuaweiAdsResp2BidderResp: openRTBRequest.imp is nil")
+	}
+	if huaweiAdsResponse.Multiad == nil {
+		return nil, errors.New("convertHuaweiAdsResp2BidderResp: huaweiAdsResponse.Multiad is nil")
+	}
+
+	for _, ad30 := range huaweiAdsResponse.Multiad {
+		if mapSlotid2Imp[ad30.Slotid].ID == "" {
+			continue
+		}
+
+		if ad30.Retcode30 != 200 {
+			continue
+		}
```

Comment:
Please add tests for the `mapSlotid2Imp[ad30.Slotid].ID == ""` and `ad30.Retcode30 != 200` conditions.

Reply:
added bad_response_retcode30_204.json bad_response_dont_find_impid.json